presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java

```diff
@@ -38,4 +62,721 @@ public static JsonGenerator createJsonGenerator(JsonFactory factory, SliceOutput
+            {
+                if (block.isNull(position)) {
+                    jsonGenerator.writeNullField(objectKey);
+                }
+                else {
+                    long value = type.getLong(block, position);
+                    jsonGenerator.writeStringField(objectKey, printTimestampWithoutTimeZone(session.getTimeZoneKey(), value));
+                }
+            }
+        }
+
+        private static class JsonGeneratorArrayWriter
+                extends JsonGeneratorWriter
+        {
+            private ArrayType type;
+            private JsonGeneratorWriter elementWriter;
+
+            public JsonGeneratorArrayWriter(ArrayType type)
+            {
+                this.type = type;
+                this.elementWriter = createJsonGeneratorWriter(type.getElementType());
+            }
+
+            @Override
+            public void writeJsonValue(JsonGenerator jsonGenerator, Block block, int position, ConnectorSession session)
+                    throws IOException
+            {
+                if (block.isNull(position)) {
+                    jsonGenerator.writeNull();
+                }
+                else {
+                    Block arrayBlock = type.getObject(block, position);
+                    jsonGenerator.writeStartArray();
+                    for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
+                        elementWriter.writeJsonValue(jsonGenerator, arrayBlock, i, session);
+                    }
+                    jsonGenerator.writeEndArray();
+                }
+            }
+
+            @Override
+            public void writeObjectKeyValuePair(JsonGenerator jsonGenerator, String objectKey, Block block, int position, ConnectorSession session)
+                    throws IOException
+            {
+                if (block.isNull(position)) {
+                    jsonGenerator.writeNullField(objectKey);
+                }
+                else {
+                    Block arrayBlock = type.getObject(block, position);
+                    jsonGenerator.writeFieldName(objectKey);
+                    jsonGenerator.writeStartArray();
+                    for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
+                        elementWriter.writeJsonValue(jsonGenerator, arrayBlock, i, session);
+                    }
+                    jsonGenerator.writeEndArray();
+                }
+            }
+        }
+
+        private static class JsonGeneratorMapWriter
+                extends JsonGeneratorWriter
+        {
+            private MapType type;
+            private MapObjectKeyReader keyReader;
+            private JsonGeneratorWriter valueWriter;
+
+            public JsonGeneratorMapWriter(MapType type)
+            {
+                this.type = type;
+                this.keyReader = MapObjectKeyReader.createMapObjectKeyReader(type.getKeyType());
+                this.valueWriter = createJsonGeneratorWriter(type.getValueType());
```

Comment:
I would do this in `createJsonGeneratorWriter` and pass in a `JsonGeneratorWriter` instead. ditto for array and row.